extensions/native/circuit/src/multi_observe/mod.rs

```diff
@@ -0,0 +1,694 @@
+                &read_init_pos,
+            )
+            .eval(builder, is_first);
+        
+        self.memory_bridge
+            .read(
+                MemoryAddress::new(self.address_space, input_register_3),
+                [len],
+                first_timestamp + AB::F::from_canonical_usize(3),
+                &read_len,
+            )
+            .eval(builder, is_first);
+
+        self.memory_bridge
+            .read(
+                MemoryAddress::new(self.address_space, input_ptr + counter - AB::F::ONE),
+                [data],
+                first_timestamp + curr_timestamp,
+                &read_data
+            )
+            .eval(builder, is_observe);
+            
+        self.memory_bridge
+            .write(
+                MemoryAddress::new(
+                    self.address_space,
+                    state_ptr + state_idx,
+                ),
+                [data],
+                first_timestamp + curr_timestamp + AB::F::ONE,
+                &write_data,
+            )
+            .eval(builder, is_observe);
+
+        self.memory_bridge
+            .read(
+                MemoryAddress::new(
+                    self.address_space,
+                    state_ptr,
+                ),
+                permutation_input,
+                first_timestamp + curr_timestamp + AB::F::TWO,
+                &read_sponge_state,
+            )
+            .eval(builder, should_permute);
+        
+        self.memory_bridge
+            .write(
+                MemoryAddress::new(
+                    self.address_space,
+                    state_ptr
+                ),
+                permutation_output,
+                first_timestamp + curr_timestamp + AB::F::from_canonical_usize(3),
+                &write_sponge_state,
+            )
+            .eval(builder, should_permute);
+
+        self.memory_bridge
+            .write(
+                MemoryAddress::new(
+                    self.address_space,
+                    input_register_1,
+                ),
+                [final_idx],
+                first_timestamp + curr_timestamp + is_first * AB::F::from_canonical_usize(4) + is_observe * AB::F::TWO + should_permute * AB::F::TWO,
+                &write_final_idx
+            )
+            .eval(builder, is_final);
+
+        // Binary indicators columns
+        builder.assert_bool(enable);
+        builder.assert_bool(is_first);
+        builder.assert_bool(is_observe);
+        builder.assert_bool(is_final);
+        builder.assert_bool(should_permute);
+
+        // Except header rows, any other rows must observe an element
+        // All rows following the header row must observe an element
+        builder
+            .when(enable)
+            .assert_eq(is_first + is_observe, AB::F::ONE);
+        builder
+            .when(is_observe)
+            .when(next.enable)
+            .when(not(next.is_first))
+            .assert_eq(next.is_observe, AB::F::ONE);
```

Comment:
It will be nice if we can reduce this constraint's degree to 3, then we can make configuration like `log_blowup = 1` to work.